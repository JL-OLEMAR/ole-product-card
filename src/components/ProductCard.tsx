import React, { createContext } from 'react';

import { useProduct } from '../hooks/useProduct';
import {
  ProductCardProps,
  ProductContextProps,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

// Here, create a context that will be used by ProductCard.
export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

// Component Main ProductCard, and Provider to the children
export function ProductCard({
  className,
  style,
  product,
  value,
  initialValues,
  children,
  onChange,
}: ProductCardProps) {
  const {
    counter,
    maxCount,
    isMaxCountReached,
    increaseBy,
    reset,
  } = useProduct({ product, value, initialValues, onChange });

  return (
    <Provider value={{ counter, maxCount, product, increaseBy }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          product,
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
}
