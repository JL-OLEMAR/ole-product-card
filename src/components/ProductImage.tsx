import React, { useContext } from 'react'

import { ProductImageProps } from '../interfaces/interfaces'
import noImage from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'

import { ProductContext } from './ProductCard'

export function ProductImage({ img, className, style }: ProductImageProps) {
  const { product } = useContext(ProductContext)
  let imgToShow: string

  if (img) {
    imgToShow = img
  } else if (product.img) {
    imgToShow = product.img
  } else {
    imgToShow = noImage
  }

  return (
    <img
      alt='Product image'
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      style={style}
    />
  )
}
