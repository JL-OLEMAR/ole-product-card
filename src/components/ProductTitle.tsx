import React, { useContext } from 'react'

import { ProductTitleProps } from '../interfaces/interfaces'
import styles from '../styles/styles.module.css'

import { ProductContext } from './ProductCard'

export function ProductTitle({ title, className, style }: ProductTitleProps) {
  const { product } = useContext(ProductContext)

  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title ? title : product.title}
    </span>
  )
}
