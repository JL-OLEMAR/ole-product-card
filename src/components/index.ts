import { ProductCardHOCProps } from '../interfaces/interfaces'

import { ProductCard as ProductCardHOC } from './ProductCard'
import { ProductButtons } from './ProductButtons'
import { ProductImage } from './ProductImage'
import { ProductTitle } from './ProductTitle'

export { ProductButtons } from './ProductButtons'
export { ProductImage } from './ProductImage'
export { ProductTitle } from './ProductTitle'

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Buttons: ProductButtons,
  Image: ProductImage,
  Title: ProductTitle
})

export default ProductCard
