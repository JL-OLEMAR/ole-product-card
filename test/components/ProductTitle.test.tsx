import * as React from 'react';
import renderer from 'react-test-renderer';

import { product1 } from '../data/products';
import { ProductCard, ProductTitle } from '../../src/components';

describe('ProductTitle', () => {
  test('should display the component correctly with the custom title', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Custom Title" className="custom-class" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should display the component with the product name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
