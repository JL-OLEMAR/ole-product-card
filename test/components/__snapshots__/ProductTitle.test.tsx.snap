// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should display the component correctly with the custom title 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Title
</span>
`;

exports[`ProductTitle should display the component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Without image
  </span>
</div>
`;
